# E sobre as profissões que devem desaparecer?

Ouvimos tanto falar disso que já passamos do ponto de acreditar, mas é melhor não relaxarmos muito. Trago aqui um exemplo sobre revisão de textos.

Em 2016 fiz minha primeira grande viagem de bicicleta, foram aproximadamente 800 km em 8 dias, com uma bicicleta da década de 50 comprada no leboncoin.com, equivalente ao olx da França. Eu vivia por lá, bem no centro do Vale do Loire, e resolvi me atirar com a velha bike e uma barraquinha para poder sentir que realmente aproveitei e conheci a região a fundo.

Mais tarde, organizando as milhares de fotos e as postagens que fiz ao longo da viagem, resolvi compilar tudo em um foto-livro, daqueles que você monta on-line e impreme em tiragens de 1x. O que descobri rapidamente é que gostar de escrever não te faz ser bom nisso. Na época usei o corretor ortogáfico do meu processador de texto para a gramática mais básica, e pedi para meus pais, que também não são grandes coisa em nosso idioma, revisarem quanto ao estilo e erros que os corretores não viam naquela época (tipo concordância, frases confusas, etc). Cada um revisou por separado e ficou ó, nota 2. Quando dois anos depois, minha namorada jornalista, a primeira pessoa com verdadeira intimidade com o idioma, pegou no livro pela primeira vez, notou um erro grosseiro logo no índice. Vergonhoso. Não sabia onde enfiar a cara.
<p align="center">
  <img width="30%" src="028 - Capa.jpg">
</p>
Desde então tenho uma coceira grande para contratar um revisor profissional que corrija e ajuste o texto para mim, para então fazer uma editoração mais pró e testar meu talento de escritor no mercado. Mas nunca mais consegui juntar tempo ($$) e energia ($$$) para fazer isso, pois tudo isso custa, e tenho em mente que não é o tipo de projeto que se paga sem muita, muita dedicação. Não está em meus planos.

## Mas foi aí que um tal de GPT entrou em minha vida

Já passamos por aquele ponto quando um computador é capaz de escrever melhor que um ser humano médio. Duas ressalvas quanto as essa afirmação:  
1. "Um computador" é modo de dizer, sabemos que não estamos falando de apenas "um computador";
2. Gabriel Garcia Marques e outros do tipo, não são seres humanos médios. Eu sou um ser humano médio.

Voltando ao ponto, ainda não estamos discutindo se um robô é capaz de criar textos tão belos quanto alguns pouco de nós somos capazes de criuar - mesmo por quê um texto "belo" precisa de talento tanto para ser criado quanto para ser lido e reconhecido como tal - mas já estamos discutindo a capacidade de um robô reconhecer erros de estilo, de concordância e de gramática, tão bem quanto um ser humano médio. Tão bem quanto mas mais do que isso, muito mais rapidamente.

Resolvi que ao invéz de encontrar (e pargar por) um revisor, trablharia com o que aprendi recentemente sobre IA para fazer uma revisão de meu livro usando um LLM. O prompt (solicitação básica) é muito simples:

In [48]:
SYSTEM_PROMPT = "Você é um revisor de textos. Por favor revisar o texto fornecido pelo usuário, em português do brasil, corrigindo gramática, estilo e forma, deixando o texto fácil de entender mas preservando as ideias e mensagens originais."

Daí é só pegar o texto completo em formato '.txt', por exemplo, e ir revisando parágrafo a parágrafo. Mas precisaremos pensar também em hiperparâmetros e em outros detalhes que só descubriremos testando.

Vou trabalhar com o API da OpenAI qué onde tenho um pouco mais de familiaridade:

In [ ]:
# Instalando a biblioteca
!pip install openai

In [2]:
#importar biblioteca
import openai

# Assignar chave
OPENAI_API_KEY = "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"
openai.api_key = OPENAI_API_KEY 

Para conseguir uma chave, basta ir no site da openai (https://openai.com/), fazer um login e na área de API Keys, criar quantas você quiser.

Vamos abrir o arquivo texto e testar com um dos parágrafos iniciais:

In [3]:
# Abrir arquivo "Retiro Espiritual.txt" na lista texto separando por parágrafos (quebra de linha)
with open("Retiro Espiritual.txt", 'r', encoding='utf-8') as f:
    texto = f.read().split('\n')

# Visualizar 5 primeiras linhas incluindo o número de palavas e de caracteres por parágrafo
for i, p in enumerate(texto[0:5]):
    print(i,'words:', len(p.split()),'chars:', len(p),':', p)

0 words: 1 chars: 11 : #INTRODUÇÃO
1 words: 15 chars: 73 : É importante começar explicando o que é e o que não é este pequeno livro.
2 words: 62 chars: 368 : Este livro não se trata de um guia de viagens. Guias de viagem têm por objetivo guiar o viajante dando-lhe opções de trajetos, pontos a visitar, um pouco de história e contexto para que cada local possa ser relativamente compreendido e admirado, e informações práticas sobre onde se hospedar, onde comer, onde comprar, etc. Este livro nem sequer tenta fazer tudo isso.
3 words: 165 chars: 1050 : O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, porém convidam o viajante visualmente a experimentar o local dando-lhe uma idéia de o que ele vai encontrar, resolvendo expectativas. Mas isso nem sempre funciona. Fotos revelam apenas uma perspectiva particular de um momento particular, e o fotógrafo, muitas vezes, tem como objetivo a busca 

Vejam que para facilitar minha vida e automatizar a coisa toda facilmente, eliminei linhas em branco e coloquei '#' para identificar títulos de parágrafos e seções.

In [4]:
# Como estamos lidando com textos longos, vamos usar uma ferramenta melhor para impressão:
from IPython.display import display, Markdown

Vamos então fazer uma primeira chamanda em nosso GPT com o primeiro parágrafo 'texto[1]'

In [49]:
# Fazer um chamado padrão com a primeira linha (texto[1]) usando o modelo gpt-3.5-turbo que,
# ao menos enquanto vos escrevo, é o melhor custo-benefício.
paragrafo = texto[1]
display(Markdown('Original: ' + paragrafo))
completion = openai.chat.completions.create(
                    model="gpt-3.5-turbo",
                    messages = [
                        { "role": "system", "content": SYSTEM_PROMPT },
                        { "role": "user", "content": paragrafo }]
                    )

# Imprimir mensagem de retorno
display(Markdown('Revisão : ' + completion.choices[0].message.content))

Original: É importante começar explicando o que é e o que não é este pequeno livro.

Revisão : É importante explicar o que é e o que não é este pequeno livro.

Notamos aqui duas coisas:
1. O modelo levou apenas 2.6 segundos para me revisar uma pequena frase de 15 palavras. Queria super-humano, já o temos!
2. Notamos que fez pequenas alterações na frase, melhorando o estilo. Eu achei que ficou bom.

Vamos testar com o terceiro parágrafo, que contém 165 palavras.

In [50]:
paragrafo = texto[3]

completion = openai.chat.completions.create(
                    model="gpt-3.5-turbo",
                    messages = [
                        { "role": "system", "content": SYSTEM_PROMPT },
                        { "role": "user", "content": paragrafo }]
                    )
texto_2 = completion.choices[0].message.content

display(Markdown('Original: ' + paragrafo))
display(Markdown('Revisão : ' + texto_2))

Original: O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, porém convidam o viajante visualmente a experimentar o local dando-lhe uma idéia de o que ele vai encontrar, resolvendo expectativas. Mas isso nem sempre funciona. Fotos revelam apenas uma perspectiva particular de um momento particular, e o fotógrafo, muitas vezes, tem como objetivo a busca da beleza estética, mesmo em lugares onde é difícil encontrá-la. Como não tento cobrir uma região por completo nem tento retratar a realidade completa do local, poderia tentar classificar o livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional nem carreguei comigo durante a viagem equipamento que faça justiça a um livro fotográfico de qualidade. As fotos aqui foram feitas por um amador, com equipamento amador: um telefone celular e uma maquininha de bolso.

Revisão : O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, mas convidam visualmente o viajante a experimentar o local e dão uma ideia do que ele vai encontrar, correspondendo às expectativas. No entanto, nem sempre isso funciona. As fotos revelam apenas uma perspectiva particular de um momento específico, e muitas vezes o fotógrafo busca a beleza estética, mesmo em lugares onde é difícil encontrá-la. Como não tentei cobrir completamente uma região nem retratar a realidade completa do local, posso tentar classificar o livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que, apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional e não levei equipamentos que fizessem justiça a um livro fotográfico de qualidade durante a viagem. As fotos aqui foram feitas por um amador, com equipamento amador: um telefone celular e uma câmera de bolso.

Agora levamos 35 segundos pata revisar um parágrafo inteiro. Gostei. Vamos rodar as primeiras dez linhas:

In [51]:
# Primeiro vamos definir uma função com o chamado do GPT
def chamadaGPT(texto):
    completion = openai.chat.completions.create(
                    model="gpt-3.5-turbo",
                    messages = [
                        { "role": "system", "content": SYSTEM_PROMPT },
                        { "role": "user", "content": texto }]
                    )
    return completion.choices[0].message.content

#Acrescentando o TQDM, o melhor antídoto para ansiedade (barra de progresso)
from tqdm.auto import tqdm

# Agora vamos parágrafo a parágrafo até o parágrafo 10, só para fazer um teste em escala

rev = list()
for p in tqdm(texto[0:10]):
    rev.append(chamadaGPT(p))

# re montando a lista no forato de parágrafos
rev = '\n'.join(rev)
# Guardando em arquivo txt
with open('Teste_Primeiro_Capitulo.txt', 'w', encoding='utf-8') as f:
    f.write(rev)

  0%|          | 0/10 [00:00<?, ?it/s]

Notem que ele levou apenas 7 minutos para revisar 10 parágrafos.
Vamos dar uma boa olhada no que foi feito mostrando os dois textos lado a lado.

In [52]:
# Primeiramente criar uma funçãozinha para facilitar o trabalho de mostrar os textos lado a lado
def lado_a_lado(texto1, texto2):
    #Acertar quebras de linha para MD
    texto1 = texto1.replace('\n','</p><p>')    
    texto2 = texto2.replace('\n','</p><p>')
    # Criar string MD
    markdown_string = f"""
<table style="width: 100%; background-color: gray; border-collapse: collapse;"><tr>
<td style="border: 1px solid transparent; text-align: center;">
Original
</td><td style="border: 1px solid transparent; text-align: center;">
Revisão GPT
</td></tr>
<tr><td style="border: 1px solid transparent; text-align: left;">
{texto1}
</td><td style="border: 1px solid transparent; text-align: left;">
{texto2}
</td></tr>
</table>
"""
    # mostrar a MD
    display(Markdown(markdown_string))

In [53]:
# Abrindo o arquivo recém criado
with open("Teste_Primeiro_Capitulo.txt", 'r', encoding='utf-8') as g:
    rev = g.read()

texto1 = '\n'.join(texto[0:10])
texto2 = rev

# Função para mostar texto lado a lado
lado_a_lado(texto1, texto2)


<table style="width: 100%; background-color: gray; border-collapse: collapse;"><tr>
<td style="border: 1px solid transparent; text-align: center;">
Original
</td><td style="border: 1px solid transparent; text-align: center;">
Revisão GPT
</td></tr>
<tr><td style="border: 1px solid transparent; text-align: left;">
#INTRODUÇÃO</p><p>É importante começar explicando o que é e o que não é este pequeno livro.</p><p>Este livro não se trata de um guia de viagens. Guias de viagem têm por objetivo guiar o viajante dando-lhe opções de trajetos, pontos a visitar, um pouco de história e contexto para que cada local possa ser relativamente compreendido e admirado, e informações práticas sobre onde se hospedar, onde comer, onde comprar, etc. Este livro nem sequer tenta fazer tudo isso.</p><p>O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, porém convidam o viajante visualmente a experimentar o local dando-lhe uma idéia de o que ele vai encontrar, resolvendo expectativas. Mas isso nem sempre funciona. Fotos revelam apenas uma perspectiva particular de um momento particular, e o fotógrafo, muitas vezes, tem como objetivo a busca da beleza estética, mesmo em lugares onde é difícil encontrá-la. Como não tento cobrir uma região por completo nem tento retratar a realidade completa do local, poderia tentar classificar o livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional nem carreguei comigo durante a viagem equipamento que faça justiça a um livro fotográfico de qualidade. As fotos aqui foram feitas por um amador, com equipamento amador: um telefone celular e uma maquininha de bolso.</p><p>Também não é um livro de ciclismo. Apesar de eu falar por experiência sobre a preparação de uma viagem de uma semana em bicicleta, de alguns de meus erros e acertos, e de pequenas dicas que encontrei durante todo o processo, não tenho nenhum objetivo didático nem compreensivo no assunto. Não se trata tampouco de um livro de auto-ajuda. Apesar de ser exatamente o que eu buscava nessa aventura, auto ajudar-me,  não procuro dar conselhos mágicos sobre como qualquer um pode atingir objetivos pessoais ou profissionais com meias verdades ou obviedades fantasiadas de filosofias vazias e frases de efeito sem conteúdo. Menos ainda um guia de vinhos, falo algo sobre a região vinícola por onde passo e sobre os vinhos que bebi com certa paixão, mas longe, longe de ser um expert ou referência no assunto.</p><p>O livro é nada mais que um álbum de fotografias misturado com textos adaptados de meu blog onde tentei contar em tempo real pelo que eu estava passando e o que eu estava fazendo. Meu blog “doRabuja” (dorabuja.blogspot.com) é uma de minhas ferramentas de alívio de tensão, onde solto por vezes meus cachorros, preservando minha familia e meus amigos dos malefícios de minha rabujice. Para não estar constantemente reclamando na orelha de gente que me quer bem e me suporta de diversas maneiras, escrevo, para que apenas os que realmente querem saber o que passa em minha estranha cabeça possam fazê-lo livremente, sem pressão, nos momentos que mais lhes convirem. Textos foram corrigidos (blogar a partir de um telefone celular tem suas restrições, principalmente aos com mais de 40 anos), melhorados, enriquecidos ou adaptados, mas a base sempre foi o blog.</p><p>Se for para recomendar o livro para alguém, fora de meu círculo familiar e de amigos mais próximos, recomendo aos que gostam de andar de bicicleta, de viajar, de apreciar o planeta e seus frutos, e têm algum interesse em conhecer a região do Vale do Loire. Não pretendi em nenhum momento escrever uma boa introdução a todos estes assuntos, mas estou seguro que o compartilhamento de toda a experiência é válido. Ignorem, divirtam-se ou pensem a respeito dos poucos momentos em que faço digressões espirituais mais sérias, afinal não existe ainda disponível no mercado, de forma tão completa, outro, que não o nosso próprio, singular, limitado e ilusionado ponto de vista.</p><p>#PREFÁCIOS</p><p>##La Loire, quem sou eu e onde estou agora mesmo - 30/08/2016</p><p>"La Loire" é o rio mais longo da França. Nasce na comunidade de Sainte-Eulalie, aos pés do monte Gerbier de Jonc, a 1400 metros de altitude e pouco mais de 1000 quilômetros depois, desemboca no Golfo da Gasconha, oceano Atlântico. Geograficamente é dividido em três sub-bacias hidrográficas, Loire Superior, de suas nascentes até Cuffi, onde recebe seu afluente Allier, Loire Médio, de Cuffi até Candes-Saint-Martin onde se junta às águas do Vienne, e Baixo Loire, daí até o estuário no oceano Atlântico.
</td><td style="border: 1px solid transparent; text-align: left;">
A importância da proteção ambiental se tornou uma preocupação crescente em todo o mundo. À medida que a sociedade evolui, as ações humanas têm gerado impactos negativos significativos no meio ambiente. Portanto, é fundamental promover a conscientização sobre a preservação dos recursos naturais e incentivar práticas sustentáveis ​​em todos os setores.</p><p></p><p>#COMO AGIR NA PROTEÇÃO AMBIENTAL</p><p></p><p>Existem várias maneiras pelas quais podemos contribuir para a proteção ambiental em nosso dia a dia. Realizando pequenas mudanças em nossos hábitos diários, podemos fazer a diferença. Alguns exemplos práticos incluem economizar água e energia, reduzir, reutilizar e reciclar produtos, utilizar transportes públicos e compartilhar carros, plantar árvores e utilizar produtos sustentáveis.</p><p></p><p>Para economizar água e energia, devemos fechar a torneira enquanto escovamos os dentes, tomar banhos mais curtos e utilizar lâmpadas de LED ou fluorescentes em nossas residências.</p><p></p><p>A redução, reutilização e reciclagem de produtos é uma forma importante de evitar o desperdício de recursos naturais e a poluição. Devemos optar por produtos duráveis ​​e recicláveis, além de reduzir o consumo de plástico descartável e embalagens desnecessárias.</p><p></p><p>A utilização de transportes públicos e o compartilhamento de carros ajudam a reduzir a emissão de gases poluentes, além de contribuir para um trânsito mais fluído. Alternativas como bicicleta ou caminhar também são ótimas opções em pequenas distâncias.</p><p></p><p>Plantar árvores, além de melhorar o ar que respiramos, ajuda a reduzir o aquecimento global, já que as árvores absorvem dióxido de carbono e liberam oxigênio. Participar de projetos de reflorestamento e apoiar iniciativas de arborização urbana são excelentes maneiras de contribuir para a proteção ambiental.</p><p></p><p>Além disso, devemos optar por produtos sustentáveis ​​e ecologicamente corretos. Ao escolher alimentos orgânicos, produtos de limpeza biodegradáveis ​​e roupas de algodão orgânico, estamos promovendo práticas sustentáveis ​​e apoiando a preservação do meio ambiente.</p><p></p><p>#CONCLUSÃO</p><p></p><p>A proteção ambiental é responsabilidade de todos nós. Pequenas atitudes em nosso cotidiano podem ter um grande impacto no meio ambiente. Portanto, devemos estar cientes da importância de preservar os recursos naturais e adotar práticas sustentáveis ​​em todas as áreas da nossa vida. Somente através de uma mudança de mentalidade e ação coletiva poderemos garantir um futuro ambientalmente saudável para as próximas gerações.</p><p>É importante começar explicando o que este pequeno livro é e o que ele não é.</p><p>Este livro não é um guia de viagens. Os guias de viagem têm como objetivo orientar os viajantes, fornecendo opções de trajetos, pontos turísticos para visitar, um pouco de história e contexto para que cada local possa ser compreendido e apreciado, além de informações práticas sobre hospedagem, restaurantes, compras, entre outros. No entanto, este livro não pretende fazer tudo isso.</p><p>O livro não pode ser classificado como um guia fotográfico. Guias fotográficos geralmente possuem menos informações práticas e histórias, mas convidam visualmente os viajantes a experimentar o local, dando-lhes uma ideia do que esperar e atendendo às suas expectativas. No entanto, isso nem sempre funciona. As fotos capturam apenas uma perspectiva particular de um momento específico e muitas vezes visam buscar a beleza estética, mesmo em lugares onde ela é difícil de encontrar. Como meu objetivo não é cobrir uma região por completo nem retratar completamente a realidade do local, eu poderia classificar meu livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que, apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional e não levei equipamentos que fariam jus a um livro fotográfico de qualidade durante minha viagem. As fotos aqui foram tiradas por um amador, com equipamento amador: um telefone celular e uma câmera de bolso.</p><p>Também não é um livro sobre ciclismo. Embora eu fale por experiência sobre a preparação de uma viagem de uma semana de bicicleta, compartilhando alguns erros e acertos, e dando pequenas dicas que descobri ao longo do processo, não tenho o objetivo de ensinar ou instruir sobre o assunto. Também não se trata de um livro de autoajuda. Embora seja exatamente o que eu buscava nessa aventura, que é me ajudar, não pretendo dar conselhos milagrosos sobre como qualquer um pode alcançar metas pessoais ou profissionais com meias verdades ou frases vazias e sem conteúdo. E definitivamente não é um guia de vinhos. Eu falo um pouco sobre a região vinícola por onde passei e sobre os vinhos que provei com certo entusiasmo, mas estou longe de ser um especialista no assunto.</p><p>O livro não passa de um álbum de fotografias misturado com textos adaptados do meu blog, onde tentei contar em tempo real o que eu estava passando e o que eu estava fazendo. Meu blog, chamado “doRabuja” (dorabuja.blogspot.com), é uma das minhas ferramentas para aliviar a tensão, onde desabafo às vezes, sem prejudicar minha família e meus amigos com minha irritação. Ao invés de reclamar constantemente no ouvido das pessoas que me querem bem e me apoiam de várias maneiras, eu escrevo, para que apenas aqueles que realmente querem saber o que se passa na minha mente estranha possam fazê-lo livremente, sem pressão, nos momentos que forem mais convenientes para eles. Os textos foram corrigidos (blogar a partir de um telefone celular tem suas limitações, especialmente para quem tem mais de 40 anos), melhorados, enriquecidos ou adaptados, mas a base sempre foi o blog.</p><p>Se for recomendar o livro para alguém fora do meu círculo familiar e amigos mais próximos, eu o recomendaria para aqueles que gostam de andar de bicicleta, viajar, apreciar o planeta e suas maravilhas, e também têm interesse em conhecer a região do Vale do Loire. Não pretendi escrever uma introdução completa sobre todos esses assuntos, mas acredito que compartilhar toda a experiência é válido. Podem ignorar, se divertir ou refletir sobre os momentos em que faço reflexões mais sérias. Afinal, no mercado não há ainda outra perspectiva tão completa, além da nossa própria, singular, limitada e ilusória visão.</p><p>É importante ressaltar que não há um "certo" ou "errado" absoluto quando se trata de prefácios. Eles podem variar de acordo com o estilo do autor e o propósito do livro. No entanto, existem algumas diretrizes que podem ajudar a criar um prefácio eficaz e envolvente.</p><p></p><p>Primeiro, é essencial que o prefácio seja conciso e objetivo. Ele deve fornecer uma breve introdução ao livro, destacando os principais temas e propósitos. Evite divagações ou detalhes desnecessários que possam distrair o leitor.</p><p></p><p>Além disso, certifique-se de que o prefácio seja escrito de forma clara e coerente. Utilize frases e parágrafos bem estruturados, evitando a ambiguidade ou a confusão. O objetivo é transmitir as informações de forma clara e acessível.</p><p></p><p>Outro aspecto importante é a personalização do prefácio. Isso significa que o autor deve compartilhar suas próprias experiências e perspectivas em relação ao livro. Isso ajuda a estabelecer uma conexão emocional com o leitor e a fornecer um contexto mais amplo para a obra.</p><p></p><p>Por fim, verifique a ortografia e a gramática do prefácio. Erros gramaticais ou ortográficos podem distrarir o leitor e prejudicar a credibilidade do livro. Portanto, revise com cuidado e, se possível, peça a opinião de outros revisores ou editores antes de finalizar o texto.</p><p></p><p>Em suma, um prefácio é uma oportunidade para o autor se conectar com o leitor, transmitir a essência do livro e estabelecer o tom para a leitura. Ao seguir essas diretrizes, você estará no caminho certo para criar um prefácio impactante e envolvente.</p><p>A Loire, quem sou eu e onde estou agora - 30/08/2016</p><p></p><p>A Loire, um dos rios mais famosos da França, é conhecido por sua beleza natural e rica história.</p><p></p><p>Neste momento, encontro-me às margens do Rio Loire, apreciando o seu esplendor. A tranquilidade das suas águas e a exuberância da paisagem que as rodeia encantam-me. À medida que observo as árvores e as aves que sobrevoam o rio, sinto-me profundamente conectado à natureza.</p><p></p><p>A Loire é mais do que apenas um rio para mim. Ela é uma fonte de inspiração e evoca uma sensação de calma e serenidade dentro de mim. À medida que caminho ao longo das suas margens, sinto-me imerso em sua história e em toda a cultura que a rodeia.</p><p></p><p>Esta jornada ao longo do Rio Loire tem sido uma experiência enriquecedora. Tenho aprendido sobre a importância do rio para as comunidades locais e como ele tem desempenhado um papel vital na economia da região.</p><p></p><p>Ao longo do meu percurso, tenho explorado vilarejos pitorescos e testemunhado a magnificência dos castelos majestosos que se erguem ao longo das margens do rio. Cada parada revela uma nova história, um novo capítulo da rica história da Loire.</p><p></p><p>Aqui, na beira do Rio Loire, percebi o quão importante é apreciar a natureza e a herança cultural que nos rodeia. Não importa onde estejamos neste vasto mundo, é essencial valorizar e preservar esses tesouros, garantindo que eles sejam apreciados pelas futuras gerações.</p><p></p><p>Encontro-me em êxtase neste momento, imerso na beleza e na magia do Rio Loire. Esta jornada tem sido verdadeiramente significativa e inesquecível, permitindo-me conectar com a história, a natureza e a minha própria essência.</p><p></p><p>Assim, despeço-me temporariamente deste maravilhoso lugar, mas carrego comigo as memórias e a gratidão por esta experiência única ao longo das margens do Rio Loire.</p><p>O rio Loire é o mais longo da França. Ele nasce na comunidade de Sainte-Eulalie, aos pés do monte Gerbier de Jonc, a 1400 metros de altitude. Após percorrer um pouco mais de 1000 quilômetros, ele desemboca no Golfo da Gasconha, no oceano Atlântico. Geograficamente, o rio é dividido em três sub-bacias hidrográficas: Loire Superior, que vai das nascentes até Cuffi, onde recebe o seu afluente Allier; Loire Médio, que vai de Cuffi até Candes-Saint-Martin, onde se junta às águas do Vienne; e Baixo Loire, que vai dali até o estuário no oceano Atlântico.
</td></tr>
</table>


Note que além de revisar o meu texto, o GPT criou textos completamente novos, seus, especificamente quando o prompt era o nome do capítulo. Interessante - e grato pela ajuda - mas não era o que esperava.
Acredito que podemos resolver isso se não passarmos o nome do capítulo para a revisão e/ou se diminuiormos a "temperatura" do chat. 

Um dos parâmetros que podemos passar para o GPT é o parâmetro "temperature" que vai dizer ao modelo o quão criativo e livre você quer que ele seja na escolha de palavras. Ele pode variar de 0 a 2, sendo 1 o default. Temperaturas mais baixas torna o gerador de texto mais repetitivo, enquanto temperaturas mais altas o torna mais criativo, correndo risco de sofrer as chamadas "alucinações". 

Outro parâmetro que pode ser interessante é limitar o número de palavras (tokens) que o modelo pode usar em cada revisão. Isso normalmente trunca o texto gerado, mas é uma boa forma de dar "um corte" na ferramenta se ela começar a pirar muito.

Vamos re-fazer a função chamadaGPT corrigindo estes problemas:

In [55]:
def chamadaGPT(texto):
    # testar se 'texto' começa com '#' o que caracteriza os nomes de capítulos
    if texto[0] == '#':
        return texto 
    else:
        completion = openai.chat.completions.create(
                        model="gpt-3.5-turbo",
                        messages = [
                        { "role": "system", "content": SYSTEM_PROMPT },
                        { "role": "user", "content": texto }],
                        temperature=0.5,
                        )
        return completion.choices[0].message.content

Vamos testar novamente:

In [56]:
rev = list()
for p in tqdm(texto[0:10]):
    rev.append(chamadaGPT(p))
rev = "\n".join(rev)
with open('Teste_Primeiro_Capitulo.txt', 'w', encoding='utf-8') as f:
    f.write(rev)

  0%|          | 0/10 [00:00<?, ?it/s]

In [57]:
# Abrindo o arquivo recém criado
with open("Teste_Primeiro_Capitulo.txt", 'r', encoding='utf-8') as g:
    rev = g.read()

texto1 = '\n'.join(texto[0:10])
texto2 = rev

# Função para mostar texto lado a lado
lado_a_lado(texto1, texto2)


<table style="width: 100%; background-color: gray; border-collapse: collapse;"><tr>
<td style="border: 1px solid transparent; text-align: center;">
Original
</td><td style="border: 1px solid transparent; text-align: center;">
Revisão GPT
</td></tr>
<tr><td style="border: 1px solid transparent; text-align: left;">
#INTRODUÇÃO</p><p>É importante começar explicando o que é e o que não é este pequeno livro.</p><p>Este livro não se trata de um guia de viagens. Guias de viagem têm por objetivo guiar o viajante dando-lhe opções de trajetos, pontos a visitar, um pouco de história e contexto para que cada local possa ser relativamente compreendido e admirado, e informações práticas sobre onde se hospedar, onde comer, onde comprar, etc. Este livro nem sequer tenta fazer tudo isso.</p><p>O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, porém convidam o viajante visualmente a experimentar o local dando-lhe uma idéia de o que ele vai encontrar, resolvendo expectativas. Mas isso nem sempre funciona. Fotos revelam apenas uma perspectiva particular de um momento particular, e o fotógrafo, muitas vezes, tem como objetivo a busca da beleza estética, mesmo em lugares onde é difícil encontrá-la. Como não tento cobrir uma região por completo nem tento retratar a realidade completa do local, poderia tentar classificar o livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional nem carreguei comigo durante a viagem equipamento que faça justiça a um livro fotográfico de qualidade. As fotos aqui foram feitas por um amador, com equipamento amador: um telefone celular e uma maquininha de bolso.</p><p>Também não é um livro de ciclismo. Apesar de eu falar por experiência sobre a preparação de uma viagem de uma semana em bicicleta, de alguns de meus erros e acertos, e de pequenas dicas que encontrei durante todo o processo, não tenho nenhum objetivo didático nem compreensivo no assunto. Não se trata tampouco de um livro de auto-ajuda. Apesar de ser exatamente o que eu buscava nessa aventura, auto ajudar-me,  não procuro dar conselhos mágicos sobre como qualquer um pode atingir objetivos pessoais ou profissionais com meias verdades ou obviedades fantasiadas de filosofias vazias e frases de efeito sem conteúdo. Menos ainda um guia de vinhos, falo algo sobre a região vinícola por onde passo e sobre os vinhos que bebi com certa paixão, mas longe, longe de ser um expert ou referência no assunto.</p><p>O livro é nada mais que um álbum de fotografias misturado com textos adaptados de meu blog onde tentei contar em tempo real pelo que eu estava passando e o que eu estava fazendo. Meu blog “doRabuja” (dorabuja.blogspot.com) é uma de minhas ferramentas de alívio de tensão, onde solto por vezes meus cachorros, preservando minha familia e meus amigos dos malefícios de minha rabujice. Para não estar constantemente reclamando na orelha de gente que me quer bem e me suporta de diversas maneiras, escrevo, para que apenas os que realmente querem saber o que passa em minha estranha cabeça possam fazê-lo livremente, sem pressão, nos momentos que mais lhes convirem. Textos foram corrigidos (blogar a partir de um telefone celular tem suas restrições, principalmente aos com mais de 40 anos), melhorados, enriquecidos ou adaptados, mas a base sempre foi o blog.</p><p>Se for para recomendar o livro para alguém, fora de meu círculo familiar e de amigos mais próximos, recomendo aos que gostam de andar de bicicleta, de viajar, de apreciar o planeta e seus frutos, e têm algum interesse em conhecer a região do Vale do Loire. Não pretendi em nenhum momento escrever uma boa introdução a todos estes assuntos, mas estou seguro que o compartilhamento de toda a experiência é válido. Ignorem, divirtam-se ou pensem a respeito dos poucos momentos em que faço digressões espirituais mais sérias, afinal não existe ainda disponível no mercado, de forma tão completa, outro, que não o nosso próprio, singular, limitado e ilusionado ponto de vista.</p><p>#PREFÁCIOS</p><p>##La Loire, quem sou eu e onde estou agora mesmo - 30/08/2016</p><p>"La Loire" é o rio mais longo da França. Nasce na comunidade de Sainte-Eulalie, aos pés do monte Gerbier de Jonc, a 1400 metros de altitude e pouco mais de 1000 quilômetros depois, desemboca no Golfo da Gasconha, oceano Atlântico. Geograficamente é dividido em três sub-bacias hidrográficas, Loire Superior, de suas nascentes até Cuffi, onde recebe seu afluente Allier, Loire Médio, de Cuffi até Candes-Saint-Martin onde se junta às águas do Vienne, e Baixo Loire, daí até o estuário no oceano Atlântico.
</td><td style="border: 1px solid transparent; text-align: left;">
#INTRODUÇÃO</p><p>É importante começar explicando o que este pequeno livro é e o que não é.</p><p>Este livro não é um guia de viagens. Os guias de viagem têm o objetivo de orientar os viajantes, fornecendo opções de rotas, pontos turísticos para visitar, um pouco de história e contexto para que cada local possa ser compreendido e apreciado, além de informações práticas sobre hospedagem, restaurantes, compras, etc. No entanto, este livro não tenta fazer nada disso.</p><p>O livro também não pode ser classificado como um guia fotográfico. Bons guias fotográficos costumam trazer menos informações práticas e menos histórias, mas convidam o viajante visualmente a experimentar o local, dando-lhe uma ideia do que ele vai encontrar e resolvendo expectativas. No entanto, isso nem sempre funciona. Fotos revelam apenas uma perspectiva particular de um momento específico, e o fotógrafo, muitas vezes, tem como objetivo buscar a beleza estética, mesmo em lugares onde é difícil encontrá-la. Como não tentei cobrir uma região por completo nem retratar a realidade completa do local, poderia classificar o livro não como um guia fotográfico, mas apenas como um livro fotográfico. O problema é que, apesar de ser apaixonado por fotografia e pelo prazer estético em geral, não sou um fotógrafo profissional nem carreguei equipamento que faça justiça a um livro fotográfico de qualidade durante a viagem. As fotos aqui foram feitas por um amador, com equipamento amador: um telefone celular e uma câmera de bolso.</p><p>Também não é um livro sobre ciclismo. Embora eu fale por experiência sobre a preparação de uma viagem de uma semana de bicicleta, compartilhando alguns dos meus erros e acertos, e dando algumas dicas que aprendi ao longo do processo, não tenho a intenção de fornecer um guia abrangente sobre o assunto. Também não se trata de um livro de autoajuda. Embora essa aventura tenha sido uma forma de me ajudar a crescer pessoalmente, não estou buscando dar conselhos mágicos sobre como qualquer um pode alcançar seus objetivos pessoais ou profissionais com clichês vazios ou frases de efeito sem conteúdo. Também não é um guia de vinhos. Falo um pouco sobre a região vinícola por onde passei e sobre os vinhos que apreciei com paixão, mas estou longe de ser um especialista ou uma referência no assunto.</p><p>O livro é nada mais do que um álbum de fotografias misturado com textos adaptados do meu blog, onde tentei contar em tempo real o que eu estava passando e o que eu estava fazendo. Meu blog "doRabuja" (dorabuja.blogspot.com) é uma das minhas formas de aliviar a tensão, onde desabafo às vezes, preservando minha família e meus amigos dos efeitos negativos da minha irritação. Para não ficar reclamando constantemente no ouvido das pessoas que me querem bem e me apoiam de várias maneiras, eu escrevo, para que apenas aqueles que realmente querem saber o que se passa na minha cabeça estranha possam fazê-lo livremente, sem pressão, nos momentos que lhes convierem. Os textos foram corrigidos (blogar a partir de um telefone celular tem suas limitações, especialmente para aqueles com mais de 40 anos), melhorados, enriquecidos ou adaptados, mas a base sempre foi o blog.</p><p>Se eu fosse recomendar o livro para alguém fora do meu círculo familiar e amigos mais próximos, seria para aqueles que gostam de andar de bicicleta, viajar, apreciar o planeta e tudo o que ele tem a oferecer, e têm interesse em conhecer a região do Vale do Loire. Não tive a intenção de escrever uma introdução completa sobre todos esses assuntos, mas tenho certeza de que compartilhar toda a experiência é válido. Ignorem, divirtam-se ou reflitam sobre os poucos momentos em que faço digressões espirituais mais sérias, afinal, não há outro ponto de vista disponível no mercado que seja tão completo como o nosso próprio, singular, limitado e ilusionado ponto de vista.</p><p>#PREFÁCIOS</p><p>##La Loire, quem sou eu e onde estou agora mesmo - 30/08/2016</p><p>O rio "Loire" é o mais longo da França. Ele nasce na comunidade de Sainte-Eulalie, aos pés do monte Gerbier de Jonc, a uma altitude de 1400 metros. Após percorrer um pouco mais de 1000 quilômetros, o rio desemboca no Golfo da Gasconha, no oceano Atlântico. Geograficamente, o rio é dividido em três sub-bacias hidrográficas: o Alto Loire, que vai desde suas nascentes até Cuffi, onde recebe seu afluente Allier; o Médio Loire, que vai de Cuffi até Candes-Saint-Martin, onde se junta às águas do Vienne; e o Baixo Loire, que vai dali até o estuário no oceano Atlântico.
</td></tr>
</table>


Agora que estamos satisfeitos com o resultado, podemos passar o revisor por todo o texto e salvar em um arquivo definitivo:

In [58]:
rev = list()
for p in tqdm(texto):
    rev.append(chamadaGPT(p))
rev = "\n".join(rev)
with open('Revisao API GPT35-turbo.txt', 'w', encoding='utf-8') as f:
    f.write(rev)

  0%|          | 0/214 [00:00<?, ?it/s]

## Considerações finais sobre prazo e custo

Vejam ue para um texto com 84k caracteres, cerca de 15k palavras, o GPT precisou de apenas 1h e 35 minutos para uma revisão completa e de qualidade satisfatória. Em termos de custo, tudo, incluindo as repetidas iterações para teste e ajustes do códico, me custou extorsivos $0.11... isso mesmo. Onze centavos de dolar.

Então, me diz o que você acha, revisores de texto estão com seus dias contados?